topology/nhlt/intel/ssp-nhlt.c

```diff
@@ -0,0 +1,696 @@
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
+  The full GNU General Public License is included in this distribution
+  in the file called LICENSE.GPL.
+*/
+
+#include <stdint.h>
+#include <errno.h>
+#include <stdio.h>
+#include <alsa/input.h>
+#include <alsa/output.h>
+#include <alsa/conf.h>
+#include <alsa/error.h>
+#include "../nhlt.h"
+#include "intel-nhlt.h"
+#include "ssp-nhlt.h"
+#include "ssp/ssp-process.h"
+
+static int set_ssp_data(struct tplg_pre_processor *tplg_pp, snd_config_t *dai_cfg)
+{
+	const char *tdm_padding_per_slot;
+	snd_config_iterator_t i, next;
+	int default_hw_config_id = 0;
+	const char *direction = NULL;
+	const char *quirks = NULL;
+	int frame_pulse_width = 0;
+	snd_config_t *class_cfg;
+	int clks_control = 0;
+	int sample_bits = 0;
+	int bclk_delay = 0;
+	int dai_index = 0;
+	long int_val = 0;
+	int mclk_id = 0;
+	snd_config_t *n;
+	const char *id;
+	int ret;
+
+	tplg_pp_debug("set_ssp_data from topology");
+
+	/* get default values from class definition */
+	ret = snd_config_search(tplg_pp->input_cfg, "Class.Dai.SSP", &class_cfg);
+	if (ret < 0)
+		return -EINVAL;
+
+	snd_config_for_each(i, next, class_cfg) {
+
+		n = snd_config_iterator_entry(i);
+
+		if (snd_config_get_id(n, &id) < 0)  {
+			tplg_pp_debug("set_ssp_data no id found");
+			continue;
+		}
+
+		tplg_pp_debug("set_ssp_data default id %s", id);
+
+		if (!strcmp(id, "direction")) {
+			if (snd_config_get_string(n, &direction))
+				return -EINVAL;
+		}
+
+		if (!strcmp(id, "quirks")) {
+			if (snd_config_get_string(n, &quirks))
+				return -EINVAL;
+		}
+
+		if (!strcmp(id, "dai_index")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			dai_index = int_val;
+		}
+
+		if (!strcmp(id, "sample_bits")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			sample_bits = int_val;
+		}
+
+		if (!strcmp(id, "bclk_delay")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			bclk_delay = int_val;
+		}
+
+		if (!strcmp(id, "mclk_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			mclk_id = int_val;
+		}
+
+		if (!strcmp(id, "default_hw_config_id")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			default_hw_config_id = int_val;
+		}
+
+		if (!strcmp(id, "clks_control")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			clks_control = int_val;
+		}
+
+		if (!strcmp(id, "frame_pulse_width")) {
+			if (snd_config_get_integer(n, &int_val))
+				return -EINVAL;
+			frame_pulse_width = int_val;
+		}
+
+		if (!strcmp(id, "tdm_padding_per_slot")) {
+			if (snd_config_get_string(n, &tdm_padding_per_slot))
+				return -EINVAL;
+		}
+	}
```

Comment:
we can avoid all these checks above if we provide default values for the SSP DAI class or make them mandatory right?